Fixed Accurate Item Restoration Deletion issue

gnago opened the pull request:
Fixed an issue where accurate item restoration would overwrite items that the player has acquired since dying.

Accomplished by creating a list before restoration begins and saving would-be-replacing items into that list to be added to empty slots or dropped after restoration is complete.

gnago committed:
Fixed an issue where accurate item restoration would overwrite items that the player has acquired since dying.

gnago committed:
Comments

shakil-muntasir commented:
Your code improves the situation where items would completely replace others if they occupied the same inventory slot as before death. However, it currently duplicates the Minepacks backpack (either stacking them or placing an additional one in an empty slot). Could you address this so I can proceed with the merge?

Minepacks plugin URL: https://www.spigotmc.org/resources/minepacks-backpack-plugin-mc-1-7-1-21.19286/

gnago commented:
I'm not familiar with Minepacks, but I just tested with both the original 1.21.8 branch and my branch: Backpack duplication still occurs on your 1.21.8 if you move the backpack to a previously empty slot. I assume this was missed because the backpack item was kept in the same slot during testing.

The compatibility code from the 1.21.8 branch also doesn't seem to account for different configurations of Minepacks (e.g. if DropOnDeath is true, we wouldn't want the backpack to get removed from the drops)

Perhaps Minepacks Compatibility is an issue better approached in a separate PR? It might even be better to make a PR with Minepacks instead to check for duplicate backpacks and delete them if detected.

broken1arrow commented:
Better if you could in config more fine tune what for items to ignore in the deathchest. Even better if they using an NBT tag that could be checked for.

gnago commented:
Oh great point. I'll see to doing that

gnago committed:
Removed MinePacks references and refined passes

gnago committed:
Removed First Pass since this unnecessarily restricted what items could be present in the armour slots. This would prevent player/mob heads and custom items from being restored properly. Now every item and slot have the same restoration behaviour.

gnago commented:
The config code isn't setup the way I expected it to be and unfortunately I have other projects that I need to prioritize. IMO what we have here currently is the MVP for a functioning version of deadchest for 1.21.8 with the accurate item restoration enhancement. Backpack/NBT filtering is a separate issue.

I've gone ahead and removed the MinePacks references since I can't work too much more on this.

I also removed the first pass since this was largely a remnant of the inaccurate item restoration from before trying to intelligently equip armour. With the new system we do not need to perform checks such as whether the item was a helmet (which would actually make the player lose player/mob heads that were equipped when they died).

shakil-muntasir commented:
> The config code isn't setup the way I expected it to be and unfortunately I have other projects that I need to prioritize. IMO what we have here currently is the MVP for a functioning version of deadchest for 1.21.8 with the accurate item restoration enhancement. Backpack/NBT filtering is a separate issue.
> 
> I've gone ahead and removed the MinePacks references since I can't work too much more on this.
> 
> I also removed the first pass since this was largely a remnant of the inaccurate item restoration from before trying to intelligently equip armour. With the new system we do not need to perform checks such as whether the item was a helmet (which would actually make the player lose player/mob heads that were equipped when they died).


I think it should be much simpler than you’re expecting. There should only ever be one item from the MinePacks plugin per player. You could likely just add a simple `if` check to prevent the MinePacks backpack from duplicating.

gnago committed:
MinePacks compatibility: MinePack backpacks are always ignored.

gnago commented:
MinePacks compatibility has been added. MinePack backpacks will skip being added to the items to store in deadchests.

shakil-muntasir commented:
> MinePacks compatibility has been added. MinePack backpacks will skip being added to the items to store in deadchests.

I'm approving and merging this PR as author worked on the last piece of the puzzle that was left.
